integration/util/harness.go

```diff
@@ -0,0 +1,583 @@
+type ComponentsMask uint
+
+const (
+	// ComponentsRecentDataQuerying starts the distributor, kafka, query-frontend, querier, and livestores for recent data querying. (default)
+	ComponentsRecentDataQuerying ComponentsMask = 1 << iota
+	// ComponentsBackendQuerying starts the distributor, kafka, query-frontend, querier, and block-builder for backend querying.
+	ComponentsBackendQuerying
+	// ComponentsMetricsGeneration starts the distributor, kafka, metrics generator and prometheus for metrics generation.
+	ComponentsMetricsGeneration
+	// ComponentsBackendWork starts the backend scheduler and worker for testing backend work.
+	ComponentsBackendWork
+)
+
+// BackendsMask is a bitmask for controlling which object storage backends are started. These flags are meant to be used
+// with the | operator to request multiple backends from the test harness.
+// The test function will be called once for each backend and so specifying multiple backends will result in longer tests!
+// Use this only when you really need to test the behavior of Tempo against multiple backends.
+type BackendsMask uint
+
+const (
+	// BackendLocal starts the local backend. (default)
+	BackendLocal BackendsMask = 1 << iota
+	// BackendObjectStorageS3 starts the S3 backend.
+	BackendObjectStorageS3
+	// BackendObjectStorageAzure starts the Azure backend.
+	BackendObjectStorageAzure
+	// BackendObjectStorageGCS starts the GCS backend.
+	BackendObjectStorageGCS
+
+	// BackendObjectStorageAll starts all three object storage backends. A convenience flag to avoid having to specify all three backends manually.
+	BackendObjectStorageAll = BackendObjectStorageS3 | BackendObjectStorageAzure | BackendObjectStorageGCS
+)
+
+type TempoHarness struct {
+	Services     map[string]*e2e.HTTPService
+	TestScenario *e2e.Scenario
+
+	overridesPath  string
+	readinessProbe e2e.ReadinessProbe
+}
+
+// TestHarnessConfig provides configuration options for the test harness
+type TestHarnessConfig struct {
+	// ConfigOverlay is a config file that will be merged on top of config-base.yaml
+	// This allows tests to only specify the differences from the default config
+	// If empty, only config-base.yaml will be used
+	ConfigOverlay string
+
+	// ConfigTemplateData provides template variables to expand in the ConfigOverlay
+	// Template variables should use Go template syntax: {{ .VariableName }}
+	ConfigTemplateData map[string]any
+
+	// ReadinessProbe is a function to use a custom readiness probe for the Tempo services.
+	// It's used by the https tests to swap the readiness port to 3201.
+	ReadinessProbe e2e.ReadinessProbe
+
+	// PreStartHook is called before starting Tempo services to perform any setup or adjustments.
+	// It's used by a few tests to copy in required files, start prerequisite services, or adjust the template data for overlays.
+	PreStartHook func(*e2e.Scenario, map[string]any) error
+
+	// DeploymentMode specifies whether to run as single binary, microservices, or both
```

Comment:
// DeploymentMode specifies whether to run as single binary, microservices, or both

Should be

	// DeploymentMode specifies whether to run as single binary, microservices, or none

?